syntax = "proto3";

package org.sourceflow.gradient.dataset;

import "common_entity.proto";
import "code_entity.proto";

enum FeatureType {
    CONDITIONAL = 0;
    INPUT_PROPERTY = 1;
    INPUT_PARAMETER = 2;
    INPUT_RESULT = 3;
    OUTPUT_PROPERTY = 4;
    OUTPUT_PARAMETER = 5;
    OUTPUT_RESULT = 6;
}

message Feature {
    int32 element_id = 1;
    org.sourceflow.gradient.common.CanonicalName name = 2;
    org.sourceflow.gradient.code.DataType data_type = 3;
    FeatureType feature_type = 4;
    repeated int32 alias_ids = 5;
}

message FeatureDescription {
    org.sourceflow.gradient.common.UUID feature_description_id = 1;
    int32 element_id = 2;
    repeated Feature features = 4;
}

message FeatureDescriptionDetail {
    repeated FeatureDescription feature_descriptions = 1;
}

message DatasetHandleDetail {
    org.sourceflow.gradient.common.UUID dataset_id = 1;
    FeatureDescription feature_description = 2;
    repeated org.sourceflow.gradient.common.UUID datapoint_ids = 3;
}

message DatasetDetail {
    org.sourceflow.gradient.common.UUID dataset_id = 1;
    FeatureDescription feature_description = 2;
    repeated Datapoint datapoints = 3;
}

message Datapoint {
    repeated org.sourceflow.gradient.common.Datum datum = 3;
}

message DatasetMessage {
    org.sourceflow.gradient.common.ProjectContext project_context = 1;
    oneof payload {
        FeatureDescriptionDetail feature_descriptions_detail = 2;
        DatasetHandleDetail dataset_handle_detail = 3;
        DatasetDetail dataset_detail = 4;
    }
}

